public/app/features/variables/query/actions.test.ts

```diff
@@ -312,43 +298,16 @@ describe('query actions', () => {
         .whenActionIsDispatched(variablesInitTransaction({ uid: 'a uid' }))
         .whenAsyncActionIsDispatched(initQueryVariableEditor(toVariablePayload(variable)), true);
 
-      tester.thenDispatchedActionsPredicateShouldEqual((actions) => {
-        const [setDatasource, setEditor] = actions;
-        const expectedNumberOfActions = 2;
-
-        expect(setDatasource).toEqual(
-          changeVariableEditorExtended({ propName: 'dataSource', propValue: mocks['datasource'] })
-        );
-        expect(setEditor).toEqual(changeVariableEditorExtended({ propName: 'VariableQueryEditor', propValue: editor }));
-        return actions.length === expectedNumberOfActions;
-      });
-    });
-  });
-
-  describe('when initQueryVariableEditor is dispatched and variable dont have datasource', () => {
-    it('then correct actions are dispatched', async () => {
-      const variable = createVariable({ datasource: undefined });
-
-      const tester = await reduxTester<RootReducerType>()
-        .givenRootReducer(getRootReducer())
-        .whenActionIsDispatched(addVariable(toVariablePayload(variable, { global: false, index: 0, model: variable })))
-        .whenActionIsDispatched(variablesInitTransaction({ uid: 'a uid' }))
-        .whenAsyncActionIsDispatched(initQueryVariableEditor(toVariablePayload(variable)), true);
-
-      tester.thenDispatchedActionsPredicateShouldEqual((actions) => {
-        const [setDatasource] = actions;
-        const expectedNumberOfActions = 1;
-
-        expect(setDatasource).toEqual(changeVariableEditorExtended({ propName: 'dataSource', propValue: undefined }));
-        return actions.length === expectedNumberOfActions;
-      });
```

Comment:
This test currently throws a "cannot read property variables of undefined" exception in main because  the action (and functions it call) does not actually handle if the datasource is undefined.

https://github.com/grafana/grafana/blob/4b4afc7b2caf641de0d350552a74c7ee25f1abbf/public/app/features/variables/guard.ts#L143-L149

From what I can tell, none of the `getVariableQueryEditor` or `hasCustomVariableSupport` tests have a case for the datasource being undefined.

Also, `await getDataSourceSrv().get(name ?? '')` says it cannot return undefined, and it will _always_ return a datasource, so I think this test does not test an actual scenario?

Reply:
_If_ `getDataSourceSrv().get()` _can_ return undefined, we should update it's type and fix everything that breaks.